### PR TITLE
Tailored flows: Implement random patterns in LiB

### PR DESCRIPTION
**Pull request by alshakero**
#### Proposed Changes

* This imports Calypso's persisted `QueryClient` to be able to persist the design order cleanly.
* It then preserves the order using a `useQuery` hook.

#### Testing Instructions
1. Go to /setup/link-in-bio/patterns.
2. The first four designs should be random.
3. Refresh the page, they should have the same random order. 
4. Repeat incognito. Should get another order.
5. Refresh the page, they should have the same random order. 

#### Risk
This has the potential to change every query behavior in Stepper. From what I see, there is only one risky hook, which is this [one](https://github.com/Automattic/wp-calypso/blob/add/randomize-patterns-in-lib/client/landing/stepper/hooks/use-theme-details.ts#L15) written by @markbiek. Hi Mark! 👋🏼 Do you think it would be OK if this hook actually had infinite `staleTime`? If not, please let me know and I'll make it shorter/fetch afresh on every load.

### CODE REVIEW

**Comment by markbiek** [comment]
> Hi Mark! 👋🏼 Do you think it would be OK if this hook actually had infinite staleTime?

I'm not really sure of the implications of `staleTime`. When I wrote the linked hook, I was mimicking other hooks that I had seen.

I'm guessing what you're saying is fine but any additional details you can provide would help me understand better.

**Comment by alshakero** [comment]
This config

```js
{
	staleTime: Infinity,
	refetchOnWindowFocus: false,
	refetchOnReconnect: false,
}
```

Would cache the query response forever on the user's browser. If the theme details change, the user won't get the new details. Before my change, this `staleTime: infinite`, didn't mean much because things were cached forever **in memory**, so all gets forgotten after page refresh. But now it means actually forever.

**Comment by markbiek** [comment]
> Would cache the query response forever on the user's browser. If the theme details change, the user won't get the new details. Before my change, this `staleTime: infinite`, didn't mean much because things were cached forever **in memory**, so all gets forgotten after page refresh. But now it means actually forever.

Ah, that makes sense. In that case, it seems like it would be better to _not_ cache forever.

Do I need to change anything in that hook I created or will it all be taken care of in this PR?

**Comment by alshakero** [comment]
I'll take care of it 👍🏼

File: packages/pattern-picker/src/use-query-patterns.ts [hunk 1]
```diff
@@ -3,6 +3,28 @@ import { useQuery, UseQueryResult } from 'react-query';
...
+ *
+ * @param length the length of the array to order
+ * @returns an array of numbers in a random order
+ */
+function useOrder( length: number ) {
+	return useQuery(
+		[ 'ptk/order', length ],
+		async () => {
+			const indices = Array.from( { length }, ( _, i ) => i );
+			const firstFour = indices.slice( 0, 4 ).sort( () => Math.random() - 0.5 );
```
Comment (heavyweight) [inline packages/pattern-picker/src/use-query-patterns.ts:17]: Works great. Some optional feedback. Returning -1, 0 or 1 for sorting seems more readable to me
```suggestion
			const firstFour = indices.slice( 0, 4 ).sort( () => Math.floor(Math.random() * 3) - 1 );
```

**Review by heavyweight — APPROVED** [review]

File: packages/pattern-picker/src/use-query-patterns.ts (hunk 1 above)
Comment (alshakero) [reply packages/pattern-picker/src/use-query-patterns.ts:17]: This is a pretty common way to shuffle an array.

**Pull request merged**